File: src/ARJE/Utils/Python/Proxy/Packets/PacketWriter.cs
using System;
using System.IO;
using System.Text;

namespace ARJE.Utils.Python.Proxy.Packets
{
    public sealed class PacketWriter : IDisposable
    {
        public PacketWriter(Stream pipeStream)
            : this(new BinaryWriter(pipeStream, Encoding.UTF8, leaveOpen: true))
        {
        }

        private PacketWriter(BinaryWriter pipeWriter)
        {
            ArgumentNullException.ThrowIfNull(pipeWriter);

            this.PipeWriter = pipeWriter;
        }

        private BinaryWriter PipeWriter { get; }

        public void Dispose()
        {
            this.PipeWriter.Dispose();
        }

        public void WriteObject<TPacket>(TPacket packet)
            where TPacket : IOutboundProxyPacket
        {
            int id = 2000; // TODO: Get id
            int length = packet.Length;
            this.PipeWriter.Write(id);
            this.PipeWriter.Write(length);

            // TODO: Check stream pos
            packet.WriteObject(this.PipeWriter);
        }
    }
}

File: src/ARJE/PythonProxy/utils/csharp/packets/packet_writer.py
import typing
from utils.io.binary_writer import BinaryWriter
from utils.csharp.packets.outbound_proxy_packet import OutboundProxyPacket


class PacketWriter:

    def __init__(self, pipe_writer: BinaryWriter):
        self._pipe_writer: typing.Final = pipe_writer

    def write_object(self, packet: OutboundProxyPacket[typing.Any]):
        type_id: int = 202 # TODO
        packet_length: int = packet.length
        self._pipe_writer.write_int(type_id)
        self._pipe_writer.write_int(packet_length)
        packet.send_object(self._pipe_writer)
